### PR TITLE
VideoPlayer: identify DemuxStream by unique id, not by index

### PR DESCRIPTION
**Pull request by FernetMenta**
see title

@ace20022 /cc
@Jalle19 PVR addons need adjustments

### CODE REVIEW

**Comment by FernetMenta** [comment]
added pvr api bump

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp [hunk 1]
```diff
@@ -110,7 +110,12 @@ CDVDDemuxCC::~CDVDDemuxCC()
 
 CDemuxStream* CDVDDemuxCC::GetStream(int iStreamId) const
 {
-  return const_cast<CDemuxStreamSubtitle*>(&m_streams[iStreamId]);
+  for (int i=0; i<GetNrOfStreams(); i++)
```
Comment (ace20022) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp]: Indentation

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxClient.cpp [hunk 2]
```diff
@@ -327,9 +327,12 @@ DemuxPacket* CDVDDemuxClient::Read()
 
 CDemuxStream* CDVDDemuxClient::GetStream(int iStreamId) const
 {
-  if (iStreamId < 0 || iStreamId >= MAX_STREAMS)
-    return nullptr;
-  return m_streams[iStreamId];
+  for (int i=0; i<GetNrOfStreams(); i++)
```
Comment (ace20022) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxClient.cpp]: Indentation

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp]: ?

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxClient.cpp (hunk 2 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxClient.cpp]: what indentation?

**Comment by ksooo** [comment]
@FernetMenta I'm currently working on another small api change which will be ready tonight or tomorrow. Would make sense to put both your and my change into one PR. What do you think?

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp (hunk 1 above)
Comment (ace20022) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp]: In the loop init and test

**Comment by FernetMenta** [comment]
> Would make sense to put both your and my change into one PR. What do you think?

I would like to keep it separately. The pvr api change is just some side product caused by this PR. We can merge both at the same time.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp]: I don't see any wrong indentation here.

**Comment by ksooo** [comment]
> We can merge both at the same time.

okay.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp (hunk 1 above)
Comment (ace20022) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp]: Indentation is the wrong word, I mean missing spaces before and after = and <

**Comment by ksooo** [comment]
@FernetMenta for the pvr addon api bump you also need to change the version number here https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h#L80 and here https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h#L83

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp]: ah. this is intentional.

**Comment by FernetMenta** [comment]
@ksooo how should we handle this? I could drop the api bump commit and leave the bump to your pr.

**Comment by ksooo** [comment]
> I could drop the api bump commit and leave the bump to your pr.

 When do you plan to merge your pr? I'm not in a hurry with my pr and can wait to merge it until this pr got merged by you.

 Please don't forget to change api version number in xbmc_pvr_types.h as I pointed out in my last comment.

**Comment by FernetMenta** [comment]
> When do you plan to merge your pr?

asap because https://github.com/xbmc/xbmc/pull/9286 depends on this

> Please don't forget to change api version number in xbmc_pvr_types.h

done

jenkins build this please

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp (hunk 1 above)
Comment (ace20022) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp]: My auto formatter will change it next time :P

**Comment by ksooo** [comment]
@FernetMenta merging this PR asap is fine by me as my API change pr is also ready to merge.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp]: Then I will change it back :)

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by wsnipex** [comment]
sorry for the off-topic, but is there a reason the pvr api version is not handled through version.txt?

**Comment by FernetMenta** [comment]
@wsnipex I am not aware of any. Note that there is a version and a min version. For kodi.guilib at least addon.xml get generated.

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
@ksooo this PR is ready to go

**Comment by ksooo** [comment]
Okay. I will merge my PR now and directly after yours...

**Comment by ksooo** [comment]
All done. Thx.

**Comment by MilhouseVH** [comment]
@ksooo - are there any plans to update pvr.hts after this PR?

**Comment by FernetMenta** [comment]
@MilhouseVH he wouldn't have merged this without having a plan. relax :)

**Comment by fritsch** [comment]
The DemuxStrem hurt my inner eye :-)

**Comment by FernetMenta** [comment]
:) I changed the commit message but not the title of this pr

**Comment by MilhouseVH** [comment]
Testing with current Kodi master (5052ecb8b9) for Ubuntu x86_64, and playing a 3D BD ISO[1] results in Kodi looping forever while logging the following - playback of video never starts:

```
23:23:34  51.179726 T:139831669659392   DEBUG: CDVDDemuxFFmpeg::AddStream ID: 0
23:23:34  51.179802 T:139831669659392   DEBUG: CDVDDemuxFFmpeg::AddStream ID: 1
23:23:34  51.180515 T:139831669659392   DEBUG: CDVDDemuxFFmpeg::AddStream ID: 0
23:23:34  51.180618 T:139831669659392   DEBUG: CDVDDemuxFFmpeg::AddStream ID: 1
23:23:34  51.181313 T:139831669659392   DEBUG: CDVDDemuxFFmpeg::AddStream ID: 0
23:23:34  51.181404 T:139831669659392   DEBUG: CDVDDemuxFFmpeg::AddStream ID: 1
23:23:34  51.182178 T:139831669659392   DEBUG: CDVDDemuxFFmpeg::AddStream ID: 0
```
Full debug log: http://sprunge.us/SONd

1. Try the first 3D Test Clip from the [Kodi Samples](http://kodi.wiki/view/Samples) page, [Full MVC 3D ISO #1 (1080p/23.976)](https://drive.google.com/file/d/0BwxFVkl63-lEd2c0ekxfbFd3Tmc/view?usp=sharing)

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 3]
```diff
@@ -1562,10 +1549,11 @@ bool CDVDDemuxFFmpeg::IsProgramChange()
...
   {
     int idx = m_pFormatContext->programs[m_program]->stream_index[i];
-    CDemuxStream *stream = GetStreamInternal(idx);
+    int id = m_pFormatContext->streams[idx]->id;
+    CDemuxStream *stream = GetStream(id);
     if (!stream)
       return true;
-    if (m_pFormatContext->streams[idx]->codec->codec_id != stream->codec)
+    if (m_pFormatContext->streams[id]->codec->codec_id != stream->codec)
       return true;
```
Comment (ace20022) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: @FernetMenta @MilhouseVH should be idx both time and id is not needed imo.

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: thanks, I will correct it.

**Comment by MilhouseVH** [comment]
Thanks for the AudioStream logging fix, but now there is no audio at all from BD ISOs, eg. [Full MVC 3D ISO #1 (1080p/23.976)](https://drive.google.com/file/d/0BwxFVkl63-lEd2c0ekxfbFd3Tmc/view?usp=sharing) sample from the [Kodi Samples](http://kodi.wiki/view/Samples) page.

Full log: http://sprunge.us/aLKM

**Comment by ace20022** [comment]
Works for me in title mode, but not in menu mode. I'll have a look now.

**Comment by MilhouseVH** [comment]
Third time lucky... :smile:

I have a file (*.mp4) that has problems seeking after this PR - it will play normally from the start, but when seeking it will freeze once the seek is complete and will not resume playback.

Debug log, playing the file then seeking +30: http://sprunge.us/IjKZ

It might be audio related:
```
07:05:13 T:139678483556096   DEBUG: CVideoPlayer::HandleMessages - player started 2
07:05:13 T:139676815972096   DEBUG: CDVDClock::Discontinuity - CVideoPlayerAudio::OutputPacket - was:14900330.391000, should be:14837000.000000, error:-63330.391000
07:05:13 T:139676815972096   ERROR: ffmpeg[7F090AFFD700]: [aac] decode_band_types: Input buffer exhausted before END element found
07:05:13 T:139676815972096   ERROR: CVideoPlayerAudio::DecodeFrame - Decode Error. Skipping audio packet (-1094995529)
07:05:13 T:139676815972096   ERROR: ffmpeg[7F090AFFD700]: [aac] Number of bands (6) exceeds limit (2).
07:05:13 T:139676815972096   ERROR: CVideoPlayerAudio::DecodeFrame - Decode Error. Skipping audio packet (-1094995529)
07:05:13 T:139676815972096   DEBUG: ffmpeg[7F090AFFD700]: [aac] Sample rate index in program config element does not match the sample rate index configured by the container.
07:05:13 T:139676815972096   ERROR: ffmpeg[7F090AFFD700]: [aac] decode_pce: Input buffer exhausted before END element found
07:05:13 T:139676815972096   ERROR: CVideoPlayerAudio::DecodeFrame - Decode Error. Skipping audio packet (-1094995529)
07:05:13 T:139676815972096   ERROR: ffmpeg[7F090AFFD700]: [aac] channel element 3.8 is not allocated
07:05:13 T:139676815972096   ERROR: CVideoPlayerAudio::DecodeFrame - Decode Error. Skipping audio packet (-1094995529)
07:05:13 T:139676815972096   ERROR: ffmpeg[7F090AFFD700]: [aac] Reserved bit set.
07:05:13 T:139676815972096   ERROR: ffmpeg[7F090AFFD700]: [aac] Number of bands (13) exceeds limit (9).
07:05:13 T:139676815972096   ERROR: CVideoPlayerAudio::DecodeFrame - Decode Error. Skipping audio packet (-1094995529)
07:05:13 T:139676815972096   DEBUG: ffmpeg[7F090AFFD700]: [aac] Sample rate index in program config element does not match the sample rate index configured by the container.
07:05:13 T:139676815972096   ERROR: ffmpeg[7F090AFFD700]: [aac] Inconsistent channel configuration.
07:05:13 T:139676815972096   ERROR: ffmpeg[7F090AFFD700]: [aac] get_buffer() failed
07:05:13 T:139676815972096   ERROR: CVideoPlayerAudio::DecodeFrame - Decode Error. Skipping audio packet (-22)
07:05:13 T:139676815972096   ERROR: ffmpeg[7F090AFFD700]: [aac] Number of bands (9) exceeds limit (4).
07:05:13 T:139676815972096   ERROR: CVideoPlayerAudio::DecodeFrame - Decode Error. Skipping audio packet (-1094995529)
07:05:13 T:139676815972096   ERROR: ffmpeg[7F090AFFD700]: [aac] channel element 3.6 is not allocated
07:05:13 T:139676815972096   ERROR: CVideoPlayerAudio::DecodeFrame - Decode Error. Skipping audio packet (-1094995529)
```
(note that I've had to apply #9307 to get this log, as without #9307 kodi.bin will segfault upon seeking)

Prior to #9290, seeking wasn't a problem with this file. I can provide a link to the file if required (ask me on Slack).

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 4]
```diff
@@ -1611,7 +1599,7 @@ std::string CDVDDemuxFFmpeg::ConvertCodecToInternalStereoMode(const std::string
 void CDVDDemuxFFmpeg::ParsePacket(AVPacket *pkt)
 {
   AVStream *st = m_pFormatContext->streams[pkt->stream_index];
-  CDemuxStream *stream = GetStreamInternal(pkt->stream_index);
+  CDemuxStream *stream = GetStream(st->id);
```
Comment (ace20022) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: @FernetMenta @MilhouseVH Should be ```st->index```, or not?

Comment (MilhouseVH) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: Yep, that fixes it.

Comment (MilhouseVH) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: It seemed to fix it on x86, but now seems to be causing a [crash](http://sprunge.us/QbTd) on RPi - please hold off on any PR for now.

Comment (MilhouseVH) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: RPi crash is unrelated.

**Pull request merged**